SmartDeviceLink/private/SDLProtocolReceivedMessageProcessor.m

```diff
@@ -0,0 +1,317 @@
+        if (endOfMessage){
+            // Reset flag
+            endOfMessage = 0;
+            
+            // Pass the header and payload to the completion
+            messageReadyBlock(header.encrypted, header, [NSData dataWithData:self.payloadBuffer]);
+            
+            //Reset state for the next message
+            [self ResetState];
+        }
+    }
+    
+}
+
+// This is the state machine
+// It processes a single byte of a message, checks for errors,
+// and builds up a header buffer and a payload buffer
+// When the header and payload are complete, the message is processed
+// The state of the state machine effectively tracks which byte of a message we are expecting next
+// For reference: https://smartdevicelink.com/en/guides/sdl-overview-guides/protocol-spec/
+// If a byte comes in that does not conform to spec, the buffers are flushed and state is reset.
+- (Boolean)sdl_processMessagesStateMachine:(Byte)currentByte {
+    
+    Byte serviceType = 0x00;
+    Byte controlFrameInfo = 0; // "Frame Info" in the documentation
+    Boolean endOfMessageFlag = 0;
+    
+    // State determines how we process the next byte, based on previous bytes.
+    switch(state){
+        case START_STATE:
+            //Flush the buffers
+            [self ResetState];
+            
+            // 4 bits for version
+            //4 highest bits    // b1111 0000
+            version = (currentByte & 0xF0 ) >> 4;
+
+            // 1 bit for either encryption or compression, depending on version.
+            // 4th lowest bit // b0000 1000
+            encrypted = (currentByte & 0x08 ) >> 3;
+            
+            // 3 bits for frameType
+            // 3 lowest bits  // b0000 0111
+            frameType = (currentByte & 0x07 ) >> 0;
+
+            // Set the next state
+            state = SERVICE_TYPE_STATE;
+            
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Check version for errors
+            if ((version < 1 || version > 5)) {
+                prevState=state;
+                state = ERROR_STATE;
+            }
+            
+            // Check for valid frameType
+            if((frameType < SDLFrameTypeControl) || (frameType > SDLFrameTypeConsecutive)){
+                prevState=state;
+                state = ERROR_STATE;
+                break;
+            }
+            break;
+            
+        case SERVICE_TYPE_STATE:
+            // 8 bits for service type
+            serviceType = currentByte;
+
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Check for errors
+            switch (serviceType) {
+                case 0x00: //SessionType.CONTROL:
```

Comment:
```suggestion
                case SDLServiceTypeControl:
```

Do this for below cases as well

Reply:
Is there a way to simplify this?  The switch feels clunky here.  All we are trying to do is make sure the service type is one of the defined types.  It would be nice to go "if serviceType in SDLServiceTypes" or some such.